composer/utils/s3_object_store.py

```diff
@@ -0,0 +1,124 @@
+            s3_transfer_config: Optional[Dict[Any, Any]] = None) -> None:
+        self.bucket = bucket
+        config = Config(**s3_client_config)
+        self.client = boto3.client('s3', config=config)
+        self.s3_transfer_config = s3_transfer_config
+
+    def upload_object(self,
+                      file_path: str,
+                      object_name: str,
+                      bucket: Optional[str] = None,
+                      verify_hash: bool = True,
+                      extra: Optional[Dict] = None,
+                      headers: Optional[Dict[str, str]] = None):
+        """Upload an object currently located on a disk.
+
+        Args:
+            file_path (str): Path to the object on disk.
+            object_name (str): Object name (i.e. where the object will be stored in the container.)
+            verify_hash (bool, optional): Whether to verify hashes (default: ``True``)
+            extra (Optional[Dict], optional): Extra attributes to pass to the underlying provider driver.
+                (default: ``None``, which is equivalent to an empty dictionary)
+            headers (Optional[Dict[str, str]], optional): Additional request headers, such as CORS headers.
+                (defaults: ``None``, which is equivalent to an empty dictionary)
+        """
+        del headers, verify_hash # used in the libcloud interface
+
+        if bucket is None:
+            if self.bucket is None:
+                raise ValueError("No S3 bucket specified")
+            bucket = self.bucket
+
+        self.client.upload_file(Filename=file_path, Bucket=bucket, Key=object_name, ExtraArgs=extra)
+
+    def upload_object_via_stream(self,
+                                 obj: Union[bytes, Iterator[bytes]],
+                                 object_name: str,
+                                 bucket: Optional[str] = None,
+                                 extra: Optional[Dict] = None,
+                                 headers: Optional[Dict[str, str]] = None):
+        """Upload an object.
+
+        Args:
+            obj (bytes | Iterator[bytes]): The object.
+            object_name (str): Object name (i.e. where the object will be stored in the container.)
+            verify_hash (bool, optional): Whether to verify hashes (default: ``True``)
+            extra (Optional[Dict], optional): Extra attributes to pass to the underlying provider driver.
+                (default: ``None``)
+            headers (Optional[Dict[str, str]], optional): Additional request headers, such as CORS headers.
+                (defaults: ``None``)
+        """
+        del headers
+
+        if bucket is None:
+            if self.bucket is None:
+                raise ValueError("No S3 bucket specified")
+            bucket = self.bucket
+        
+        # add handling for other bytes-like types
+        if not isinstance(obj, bytes):
+            raise NotImplementedError("S3ObjectStore.upload_object_via_stream only takes a bytes object.")
```

Comment:
Can we support `iterator[bytes]`? See https://stackoverflow.com/questions/62876879/how-to-create-a-python-file-like-object-from-an-iterator